go/tao/datalog_guard.go

```diff
@@ -618,7 +624,122 @@ func (g *DatalogGuard) retract(f auth.Form) error {
 	return nil
 }
 
+func max(args ...int) int {
+	m := 0
+	for _, v := range args {
+		if m < v {
+			m = v
+		}
+	}
+	return m
+}
+
+// Figure out the max length of a principal in a term, where the length of a
+// principal is one more than the length of the extensions of the principal.
+func getMaxTermLength(t auth.Term) int {
+	if t == nil {
+		return 0
+	}
+
+	switch t.(type) {
+	case auth.Prin:
+		// TODO(tmroeder): this is not fully general, since there might be an Arg that has a longer principal.
+		return 1 + len(t.(auth.Prin).Ext)
+	case *auth.Prin:
+		// TODO(tmroeder): this is not fully general, since there might be an Arg that has a longer principal.
+		return 1 + len(t.(*auth.Prin).Ext)
+	case auth.PrinTail:
+		return len(t.(auth.PrinTail).Ext)
+	case *auth.PrinTail:
+		return len(t.(*auth.PrinTail).Ext)
+	case auth.Str, auth.Bytes, auth.Int, auth.TermVar:
+		return 0
+	default:
+		return 0
+	}
+}
+
+// Figure out the max length of a principal in a form, where the length of a
+// principal is one more than the length of the extensions of the principal.
+func getMaxFormLength(f auth.Form) int {
+	if f == nil {
+		return 0
+	}
+
+	var m []int
+	switch f.(type) {
+	case auth.Pred:
+		for _, t := range f.(auth.Pred).Arg {
+			m = append(m, getMaxTermLength(t))
```

Comment:
Make m an int, so here you can do m = max(m, getMaxTermLength(t)). There is no need to keep an array of terms just you can take the max later. Same comment several times below.